File: packages/core/src/task.ts
/**
 * Task Domain Model — DDD Aggregate Root
 *
 * Encapsulates the task lifecycle state machine with strict
 * transition rules. Every state change is validated and auditable.
 *
 * State Machine:
 *   pending → accepted → running → completed
 *                     ↘ suspended → running (resume)
 *   (any non-terminal) → cancelled | failed
 *
 * Terminal states: completed, failed, cancelled
 * Resumable states: suspended
 */

import type {
  ActiveTaskStatus,
  ContextEnvelope,
  Task,
  TaskStatus,
  TerminalTaskStatus,
  TokenBudget,
} from './types.js';

// ---------------------------------------------------------------------------
// Value Objects
// ---------------------------------------------------------------------------

/** Valid state transitions — the single source of truth */
export const TASK_TRANSITIONS: Readonly<Record<TaskStatus, readonly TaskStatus[]>> = {
  pending: ['accepted', 'cancelled', 'failed'],
  accepted: ['running', 'cancelled', 'failed'],
  running: ['completed', 'failed', 'cancelled', 'suspended'],
  completed: [],
  failed: [],
  cancelled: [],
  suspended: ['running', 'cancelled', 'failed'],
} as const;

/** States from which cancel is valid */
export const CANCELLABLE_STATES: readonly TaskStatus[] = [
  'pending',
  'accepted',
  'running',
  'suspended',
] as const;

/** States from which resume is valid */
export const RESUMABLE_STATES: readonly TaskStatus[] = ['suspended'] as const;

/** Terminal states — no further transitions */
export const TERMINAL_STATES: readonly TerminalTaskStatus[] = [
  'completed',
  'failed',
  'cancelled',
] as const;

// ---------------------------------------------------------------------------
// Domain Functions
// ---------------------------------------------------------------------------

/** Check if a transition is valid according to the state machine */
export function isValidTransition(from: TaskStatus, to: TaskStatus): boolean {
  return TASK_TRANSITIONS[from].includes(to);
}

/** Check if a status is terminal (no further transitions) */
export function isTerminal(status: TaskStatus): status is TerminalTaskStatus {
  return (TERMINAL_STATES as readonly TaskStatus[]).includes(status);
}

/** Check if a status is active (non-terminal) */
export function isActive(status: TaskStatus): status is ActiveTaskStatus {
  return !isTerminal(status);
}

/** Check if a task can be cancelled from its current state */
export function isCancellable(status: TaskStatus): boolean {
  return CANCELLABLE_STATES.includes(status);
}

/** Check if a task can be resumed from its current state */
export function isResumable(status: TaskStatus): boolean {
  return RESUMABLE_STATES.includes(status);
}

// ---------------------------------------------------------------------------
// Task Entry — Aggregate Root
// ---------------------------------------------------------------------------

/** Serializable checkpoint for task resume */
export interface TaskCheckpoint {
  /** Handler-specific state snapshot */
  data: Record<string, unknown>;
  /** When the checkpoint was created */
  created_at: number;
}

/** Immutable record of a status transition */
export interface TaskTransition {
  from: TaskStatus;
  to: TaskStatus;
  timestamp: number;
  reason?: string;
}

/**
 * TaskEntry — The aggregate root for task lifecycle.
 *
 * Encapsulates a task, its current state, abort signaling,
 * and checkpoint data for resume. All state changes go through
 * validated transitions.
 */
export interface TaskEntry {
  /** The original task definition */
  readonly task: Task;
  /** Current lifecycle status */
  status: TaskStatus;
  /** Optional context envelope */
  readonly context?: ContextEnvelope;
  /** AbortController for cancellation signaling */
  readonly abortController: AbortController;
  /** Serialized checkpoint for resume */
  checkpoint?: TaskCheckpoint;
  /** Ordered history of all transitions */
  transitions: TaskTransition[];
  /** Creation timestamp */
  readonly createdAt: number;
  /** Last update timestamp */
  updatedAt: number;
}

/**
 * Create a new TaskEntry in 'pending' state.
 * Factory function — the only way to create a valid TaskEntry.
 */
export function createTaskEntry(task: Task, context?: ContextEnvelope): TaskEntry {
  const now = Date.now();
  return {
    task,
    status: 'pending',
    context,
    abortController: new AbortController(),
    transitions: [],
    createdAt: now,
    updatedAt: now,
  };
}

/**
 * Transition a task to a new status.
 * Throws if the transition is invalid.
 * Returns the updated entry (mutates in place for performance).
 */
export function transitionTask(entry: TaskEntry, to: TaskStatus, reason?: string): TaskEntry {
  if (!isValidTransition(entry.status, to)) {
    throw new TaskTransitionError(entry.task.id, entry.status, to);
  }

  const transition: TaskTransition = {
    from: entry.status,
    to,
    timestamp: Date.now(),
    reason,
  };

  entry.transitions.push(transition);
  entry.status = to;
  entry.updatedAt = transition.timestamp;

  // Fire abort signal on cancellation
  if (to === 'cancelled' && !entry.abortController.signal.aborted) {
    entry.abortController.abort(reason ?? 'Task cancelled');
  }

  return entry;
}

/**
 * Save a checkpoint on a running task for later resume.
 */
export function saveCheckpoint(entry: TaskEntry, data: Record<string, unknown>): TaskEntry {
  if (entry.status !== 'running' && entry.status !== 'suspended') {
    throw new Error(`Cannot checkpoint task in '${entry.status}' state`);
  }
  entry.checkpoint = { data, created_at: Date.now() };
  entry.updatedAt = Date.now();
  return entry;
}

// ---------------------------------------------------------------------------
// Domain Errors
// ---------------------------------------------------------------------------

/** Thrown when an invalid state transition is attempted */
export class TaskTransitionError extends Error {
  readonly taskId: string;
  readonly from: TaskStatus;
  readonly to: TaskStatus;

  constructor(taskId: string, from: TaskStatus, to: TaskStatus) {
    super(`Invalid task transition: '${from}' → '${to}' for task '${taskId}'`);
    this.name = 'TaskTransitionError';
    this.taskId = taskId;
    this.from = from;
    this.to = to;
  }
}

File: sdk/python/src/nekte/domain/task.py
"""Task Domain Model — Aggregate Root with state machine.

State Machine:
  pending -> accepted -> running -> completed
                      -> suspended -> running (resume)
  (any non-terminal) -> cancelled | failed
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Final

from nekte.domain.types import ContextEnvelope, Task, TaskStatus

# ---------------------------------------------------------------------------
# Valid transitions — single source of truth
# ---------------------------------------------------------------------------

TASK_TRANSITIONS: Final[dict[TaskStatus, tuple[TaskStatus, ...]]] = {
    "pending": ("accepted", "cancelled", "failed"),
    "accepted": ("running", "cancelled", "failed"),
    "running": ("completed", "failed", "cancelled", "suspended"),
    "completed": (),
    "failed": (),
    "cancelled": (),
    "suspended": ("running", "cancelled", "failed"),
}

CANCELLABLE_STATES: Final[tuple[TaskStatus, ...]] = ("pending", "accepted", "running", "suspended")
RESUMABLE_STATES: Final[tuple[TaskStatus, ...]] = ("suspended",)
TERMINAL_STATES: Final[tuple[TaskStatus, ...]] = ("completed", "failed", "cancelled")


# ---------------------------------------------------------------------------
# Domain functions
# ---------------------------------------------------------------------------


def is_valid_transition(from_status: TaskStatus, to_status: TaskStatus) -> bool:
    return to_status in TASK_TRANSITIONS[from_status]


def is_terminal(status: TaskStatus) -> bool:
    return status in TERMINAL_STATES


def is_active(status: TaskStatus) -> bool:
    return not is_terminal(status)


def is_cancellable(status: TaskStatus) -> bool:
    return status in CANCELLABLE_STATES


def is_resumable(status: TaskStatus) -> bool:
    return status in RESUMABLE_STATES


# ---------------------------------------------------------------------------
# Value Objects
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class TaskCheckpoint:
    data: dict[str, Any]
    created_at: float


@dataclass(frozen=True)
class TaskTransition:
    from_status: TaskStatus
    to_status: TaskStatus
    timestamp: float
    reason: str | None = None


# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class TaskTransitionError(Exception):
    def __init__(self, task_id: str, from_status: TaskStatus, to_status: TaskStatus) -> None:
        super().__init__(
            f"Invalid task transition: '{from_status}' -> '{to_status}' for task '{task_id}'"
        )
        self.task_id = task_id
        self.from_status = from_status
        self.to_status = to_status


# ---------------------------------------------------------------------------
# TaskEntry — Aggregate Root
# ---------------------------------------------------------------------------


@dataclass
class TaskEntry:
    """Aggregate root for task lifecycle.

    Unlike the frozen Pydantic models, TaskEntry is mutable internally
    to support state transitions and checkpoint saves. External mutation
    should go through transition_task() and save_checkpoint().
    """

    task: Task
    status: TaskStatus = "pending"
    context: ContextEnvelope | None = None
    checkpoint: TaskCheckpoint | None = None
    transitions: list[TaskTransition] = field(default_factory=list)
    created_at: float = field(default_factory=time.time)
    updated_at: float = field(default_factory=time.time)
    _cancelled: bool = field(default=False, repr=False)

    @property
    def is_cancelled(self) -> bool:
        return self._cancelled


def create_task_entry(task: Task, context: ContextEnvelope | None = None) -> TaskEntry:
    """Factory — the only way to create a valid TaskEntry."""
    now = time.time()
    return TaskEntry(
        task=task,
        context=context,
        created_at=now,
        updated_at=now,
    )


def transition_task(
    entry: TaskEntry,
    to: TaskStatus,
    reason: str | None = None,
) -> TaskEntry:
    """Transition a task to a new status. Raises TaskTransitionError if invalid."""
    if not is_valid_transition(entry.status, to):
        raise TaskTransitionError(entry.task.id, entry.status, to)

    now = time.time()
    transition = TaskTransition(
        from_status=entry.status,
        to_status=to,
        timestamp=now,
        reason=reason,
    )
    entry.transitions.append(transition)
    entry.status = to
    entry.updated_at = now

    if to == "cancelled":
        entry._cancelled = True

    return entry


def save_checkpoint(entry: TaskEntry, data: dict[str, Any]) -> TaskEntry:
    """Save a checkpoint on a running/suspended task for later resume."""
    if entry.status not in ("running", "suspended"):
        raise ValueError(f"Cannot checkpoint task in '{entry.status}' state")

    entry.checkpoint = TaskCheckpoint(data=data, created_at=time.time())
    entry.updated_at = time.time()
    return entry
